Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,249 @@
+      Therefore, the kernel tends to:
+      1. reuse as much as possible a certain page that has just been freed (so called hot page, since it is stll presumably present in the HW cache)
+      2. keep for each core a stash of memory pages readily available (which prevents other cores from accessing said pages and introducing additional cache-flush operations)
+   4. The MMU, involved in performing address translations, acts also as a bus master, and performs read operations whenever it needs to do an address translation that is not already present in its own local translation cache (TLB - Translation Lookaside Buffer). Having to perform too many of such address translations (page walks) can constitute a significant performance penalty. The TLB is not very large, and accessing lots of different memory addresses that do not belong to the same translation entry can cause severe performance degradation. This is why the kernel actually keeps most of the memory mapped in the linear map, to take advantage of a feature present in many processors, that allows the mapping of large chunks of physical memory (e.g 2MB) as a single entry (or few ones). The kernel code, for example, is kept compact to maximise the efficiency of the fetching operations. For the similar reasons, the linear mapping allows to have memory already mapped, without a need for creating those mappings on the fly.
+   5. For what concerns allocation from the linear map (kmalloc / get_free_pages), the kernel attempts to keep the free pages as much continuous as possible, avoiding fragmentation of the free areas.
+      1. this is implemented through the concept of the "buddy allocator", meaning that whenever a certain amount of linear memory is requested (either sub-page or multi-page size), it always tries to obtain it from the smallest free slot available, only breaking larger free slots when no alternatives are possible.
+      2. the kernel also keeps ready a certain amount of pre-diced memory allocations, to avoid incurring the penalty of having to look for some free memory as a consequence of an allocation request.
+      3. folios are structures introduced to simplify the management of what has been traditionally called compound pages: a compound page represents a group of contiguous pages that is treated as a single logical unit. Folios could eventually support the use of optimisations provided by certain pages (e.g. ARM64 allows the use of a single page table entry to represent 16 pages, as long as they are physically contiguous and aligned to a 16 pages boundary, through the "contiguous bit" flag in the page table). This can be useful e.g. when keeping in the page cache a chunk of data from a file, should the memory be released, it could result in releasing several physically contiguous pages, instead of scattered ones.
```

Comment:
yes, but please understand that I'm not trying to explain everything that is associated with memory management, or else this will becoem a book in itself - I expect the reader to do own research on topics that might not be fully known. Or others to contribute references. The topic is complex and one who wants to work on it is expecccted to study it.

Reply:
anyways, in this specific case the purpose of the contiguous bit it to use only 1 TLB entry instead of 16 - but then I shuld explain what is a TLB and so on ...

Reply:
The thing is not that you have to explain further what the contiguous bit does, since this is clear. I just cannot see why go into that amount of detail about the implementation of folios. I think this over-complicates the whole paragraph. And then the last sentence feels detached from the beginning.
In my opinion, the important thing to know about folios is that they represent more than one physical page, and might be used to optimize memory access. But if you go deeper, and argue something is useful, then I think it is better to state clearly why, and not let the reader guessing.

Reply:
Ah, I see. Well, avoiding memory fragmentation is the goal, but I kind of took it for granted that it would stand out. I can add this.

Reply:
added